File: backend-fastapi/models/orm/place.py
from datetime import datetime
from typing import TYPE_CHECKING

from pgvector.sqlalchemy import Vector
from sqlalchemy import ForeignKey, Index, Integer, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from lib.sqlalchemy_ import BaseModel

if TYPE_CHECKING:
    from .user import User


class Place(BaseModel):
    __tablename__ = "places"
    __table_args__ = (Index("idx_place_creator", "creator_id"),)

    # Fixing mypy bug by referencing again the inherited fields
    id: Mapped[int]
    created_at: Mapped[datetime]
    updated_at: Mapped[datetime]

    # Fields
    title: Mapped[str] = mapped_column(Text, nullable=False)

    description: Mapped[str] = mapped_column(Text, nullable=False)

    address: Mapped[str] = mapped_column(Text, nullable=False)

    image_url: Mapped[str | None] = mapped_column(
        Text, nullable=False, server_default=""
    )

    location: Mapped[dict] = mapped_column(JSONB, nullable=False)

    # PgVector
    embedding: Mapped[list[float] | None] = mapped_column(
        Vector(384), nullable=True
    )

    # Relationships

    creator_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )

    creator: Mapped["User"] = relationship("User", back_populates="places")

File: backend-gin/internal/models/orm/place.go
package orm

import (
	"backend/internal/lib/gorm_"
	"database/sql/driver"
	"encoding/json"
	"errors"

	"github.com/pgvector/pgvector-go"
)

type Location struct {
	Lat float64 `json:"lat"`
	Lng float64 `json:"lng"`
}

func (l *Location) Scan(value any) error {
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("could not parse raw location vaue")
	}
	return json.Unmarshal(bytes, l)
}

func (l Location) Value() (driver.Value, error) {
	return json.Marshal(l)
}

type Place struct {
	gorm_.BaseModel

	// Fields
	Title       string           `gorm:"not null" json:"title"`
	Description string           `gorm:"not null" json:"description"`
	Address     string           `gorm:"not null" json:"address"`
	ImageURL    string           `gorm:"not null;default:''" json:"imageUrl"`
	Location    Location         `gorm:"not null;type:jsonb;serializer:json" json:"location"`
	Embedding   *pgvector.Vector `gorm:"type:vector(384)" json:"-"`

	// Foreign Keys
	CreatorID uint `gorm:"not null;index:idx_place_creator;type:integer" json:"creatorId"`
	Creator   User `gorm:"foreignKey:CreatorID;constraint:OnDelete:CASCADE" json:"creator"`
}

func (Place) TableName() string {
	return "places"
}
